svglib/svglib.py

```diff
@@ -1462,6 +1465,26 @@ def svg2rlg(path, resolve_entities=False, **kwargs):
     return drawing
 
 
+def nudge_points(points):
+    """ Nudge first coordinate if all coordinate pairs are identical.
+
+    This works around reportlab's decision to hide shapes of size zero, even
+    when the stroke should be visible.
+    """
+    if not points:
+        return
+    if len(points) < 2:
+        return
+    x0, y0 = points[:2]
+    for i in range(2, len(points)-1, 2):
```

Comment:
I wonder if this could be made any faster using built-in functions or sets...?

Reply:
Sure, I can run some performance tests. What do you think is a reasonable distribution for the number of points in a list? A standard distribution between 10 and 100?

Reply:
The best performance I found was just removing the unpacking assignments from this code. Using the `all()` function or a set comprehension were both slower.

Here's the performance test I ran:

    from timeit import timeit
    
    ITERATION_COUNT = 900000
    POINT_LISTS = []
    LIST_INDEX = 0
    
    
    def main():
        tests = ['run_test(nudge_points)',
                 'run_test(nudge_points_all)',
                 'run_test(nudge_points_set)']
        for test in tests:
            duration = timeit(test,
                              'setup()',
                              number=ITERATION_COUNT,
                              globals=globals())
            print(duration / ITERATION_COUNT, test)
    
    
    def setup():
        global LIST_INDEX
        POINT_LISTS.clear()
        for i in range(ITERATION_COUNT):
            point_count = 10 + i % 90
            points = [100, 200] * point_count
            if i % 5 != 0:
                # 4/5 of the point lists don't need a nudge.
                points[4] += 1
            POINT_LISTS.append(points)
        LIST_INDEX = 0
    
    
    def run_test(nudge):
        global LIST_INDEX
        points = POINT_LISTS[LIST_INDEX]
        nudge(points)
        LIST_INDEX += 1
    
    
    def nudge_points(points):
        if not points:
            return
        if len(points) < 4:
            return
        x = points[0]
        y = points[1]
        for i in range(2, len(points)-1, 2):
            if x != points[i] or y != points[i+1]:
                break
        else:
            # All points were identical, so we nudge.
            points[0] *= 1.0000001
    
    
    def nudge_points_all(points):
        if not points:
            return
        if len(points) < 4:
            return
        x = points[0]
        y = points[1]
        if all(points[i] == x and points[i+1] == y
               for i in range(0, len(points)-1, 2)):
            # All points were identical, so we nudge.
            points[0] *= 1.0000001
    
    
    def nudge_points_set(points):
        if not points:
            return
        if len(points) < 4:
            return
        points_set = {(points[i], points[i+1]) for i in range(0, len(points)-1, 2)}
        if len(points_set) == 1:
            # All points were identical, so we nudge.
            points[0] *= 1.0000001
    
    
    main()

The results were:

    1.6128685722247206e-06 run_test(nudge_points)
    2.4019194488856658e-06 run_test(nudge_points_all)
    7.203874138892085e-06 run_test(nudge_points_set)

Let me know if you'd like me to try any other approaches.

For now, I'll push the new version without the unpacking assignments.

Reply:
Fine then, thanks!